compiler/src/dotty/tools/dotc/transform/PickleQuotes.scala

```diff
@@ -481,10 +481,16 @@ class PickleQuotes extends MacroTransform {
         transform(tree)(using ctx.withSource(tree.source))
       else reporting.trace(i"Reifier.transform $tree at $level", show = true) {
         tree match {
-          case Apply(TypeApply(fn, (body: RefTree) :: Nil), _) if fn.symbol == defn.QuotedTypeModule_of && isCaptured(body.symbol, level + 1) =>
+          case Apply(TypeApply(fn, (body: RefTree) :: Nil), _)
+          if fn.symbol == defn.QuotedTypeModule_of && isCaptured(body.symbol, level + 1) =>
             // Optimization: avoid the full conversion when capturing `X` with `x$1: Type[X$1]`
             // in `Type.of[X]` to `Type.of[x$1.Underlying]` and go directly to `X$1`
             capturers(body.symbol)(body)
+          case Apply(Select(Apply(TypeApply(fn,_), List(ref: RefTree)),nme.apply),List(quotes))
+          if fn.symbol == defn.QuotedRuntime_exprQuote && isCaptured(ref.symbol, level + 1) =>
+            // Optimization: avoid the full conversion when capturing `x` with `x$1: Expr[X]`
+            // in `'{x}` to `'{ ${x$1} }'` and go directly to `x$1`
+            capturers(ref.symbol)(ref).select(nme.apply).appliedTo(quotes)
```

Comment:
For the example `'{ val y = $x * $x; ${ powerCode('y, n / 2) } }`, I'm wondering what's the argument to `powerCode` before and after the change.

Reply:
That was the test case I used to do this change. Unless the version of the benchmark has something else that I missed.

Reply:
Before we had

```scala
def hole1(args: Array[Any]) = 
  val y$1 = args(0).asInstanceOf[Expr[Int]]
  def hole3(args: Array[Any]) = y$1
  powerCode(unpickle('{ [[ 0 | Double ]] }, List(hole3)), n / 2)
```